pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,219 @@
+func (ipc *IPCache) upsert(endpointIP string, identity identity.NumericIdentity) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+
+	// An update is treated as a deletion and then an insert.
+	ipc.deleteLocked(endpointIP)
+
+	// Update both maps.
+	ipc.ipToIdentityCache[endpointIP] = identity
+
+	_, found := ipc.identityToIPCache[identity]
+	if !found {
+		ipc.identityToIPCache[identity] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[identity][endpointIP] = struct{}{}
+}
+
+// deleteLocked removes removes the provided IP-to-security-identity mapping
+// from both caches within ipc with the assumption that ipc's mutex is held.
+func (ipc *IPCache) deleteLocked(endpointIP string) {
+
+	identity, found := ipc.ipToIdentityCache[endpointIP]
+	if found {
+		delete(ipc.ipToIdentityCache, endpointIP)
+		delete(ipc.identityToIPCache[identity], endpointIP)
+		if len(ipc.identityToIPCache[identity]) == 0 {
+			delete(ipc.identityToIPCache, identity)
+		}
+	}
+}
+
+// delete removes the provided IP-to-security-identity mapping from both caches
+// within ipc.
+func (ipc *IPCache) delete(endpointIP string) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+	ipc.deleteLocked(endpointIP)
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	ips, exists := ipc.identityToIPCache[id]
+	return ips, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// OnIPIdentityCacheChange will be called whenever there the state of the
+	// IPCache has changed.
+	OnIPIdentityCacheChange()
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+	// see GH-2555
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
+
+	for {
+		watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+
+		// Get events from channel as they come in.
+		for event := range watcher.Events {
+
+			var (
+				cacheChanged bool
+				ipIDPair     identity.IPIdentityPair
+			)
+
+			err := json.Unmarshal(event.Value, &ipIDPair)
+			if err != nil {
+				log.Errorf("not adding entry to ip cache; error unmarshaling identity from key-value store: %s", err)
```

Comment:
From experience, it is incredibly useful to print the data that we attempted to unmarshal. In particular as it is just a bit of JSON in this case.